### PR TITLE
feat: powerbi source plugin

### PR DESCRIPTION
**Pull request by mohdsiddique**
powerbi source for Datahub "data platform"

### CODE REVIEW

File: metadata-ingestion/setup.py [hunk 1]
```diff
@@ -159,7 +162,8 @@ def get_long_description():
     "tableau": {"tableauserverclient>=0.17.0"},
     "trino": sql_common | {"trino"},
     "starburst-trino-usage": sql_common | {"trino"},
-    "nifi": {"requests", "packaging"},
+    "nifi": {"requests"},
```
Comment (shirshanka) [inline metadata-ingestion/setup.py:165]: dropping `packaging` here seems like a merge mistake?

File: metadata-ingestion/source_docs/powerbi.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+# Power Bi dashboards
```
Comment (shirshanka) [inline metadata-ingestion/source_docs/powerbi.md:1]: `s/Bi/BI`

File: metadata-ingestion/source_docs/powerbi.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+| `tenant_id`               | ✅       |                         | Power BI tenant identifier.                                                                                   |
+| `workspace_id`            | ✅       |                         | Power BI workspace identifier.                                                                                |
+| `env`                     | ✅       |                         | Environment to use in namespace when constructing URNs.                                                       |
+| `client_id`               | ✅       |                         | Azure AD App client identifier.                                                                               |
+| `client_secret`           | ✅       |                         | Azure AD App client secret.
+| `dataset_type_mapping`    | ✅       |                         | Mapping of Power BI datasource type to Datahub dataset. 
+| `scan_timeout`            | ✅       | 60                      | time in seconds to wait for Power BI metadata scan result.
+
+## Concept mapping
```
Comment (shirshanka) [inline metadata-ingestion/source_docs/powerbi.md:69]: I like this!
Are there any concepts that are not mapped? would be good to call that out.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 4]
```diff
@@ -0,0 +1,1382 @@
...
+    DATASET = "dataset"
+    DATASET_ID = "powerbi.linkedin.com/datasets/{}"
+    DATASET_KEY = "datasetKey"
+    DATASET_PROPERTIES = "datasetProperties"
+    VALUE = "value"
+    ENTITY = "ENTITY"
+    ID = "ID"
+
+
+class PowerBiAPIConfig(ConfigModel):
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:100]: extend from `EnvBasedSourceConfigBase` in `datahub.configuration.source_common`. 
That will give you the `env` field with a validator.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 5]
```diff
@@ -0,0 +1,1382 @@
...
+    # Organsation Identifier
+    tenant_id: str
+    # PowerBi workspace identifier
+    workspace_id: str
+    # Current working environment. Possible values are DEV, QA, STAGE, PROD
+    # It is used in generating dataset URN
+    env: str
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:107]: remove this

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 6]
```diff
@@ -0,0 +1,1382 @@
...
+    admin_base_url = "https://api.powerbi.com/v1.0/myorg/admin"
+    authority = "https://login.microsoftonline.com/"
+
+    def get_authority_url(self):
+        return "{}{}".format(self.authority, self.tenant_id)
+
+
+class PowerBiDashboardSourceConfig(PowerBiAPIConfig):
+    platform_name: str = "powerbi"
+    platform_urn: str = "urn:li:dataPlatform:powerbi"
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:128]: use the utilities in `datahub.utilities` to construct urns

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 7]
```diff
@@ -0,0 +1,1382 @@
...
+            return (
+                isinstance(instance, PowerBiAPI.Dataset)
+                and self.__members() == instance.__members()
+            )
+
+        def __hash__(self):
+            return hash(self.__members())
+
+    @dataclass
+    class Report:
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:223]: didn't see this in the Concept Mapping section. What is Report being mapped to?

File: metadata-ingestion/source_docs/powerbi.md (hunk 3 above)
Comment (mohdsiddique) [reply metadata-ingestion/source_docs/powerbi.md:69]: I updated the readme file

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py (hunk 6 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi.py]: Updated the code to use utility method

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py (hunk 7 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi.py:220]: Tile is either created from Report, Dataset, or Visualization. If Tile is created from Report then chart.externalUrl is set to Report.webUrl

File: metadata-ingestion/setup.py (hunk 1 above)
Comment (mohdsiddique) [reply metadata-ingestion/setup.py]: Yes, added "packaging"

File: metadata-ingestion/source_docs/powerbi.md (hunk 2 above)
Comment (mohdsiddique) [reply metadata-ingestion/source_docs/powerbi.md]: change the Bi to BI

**Comment by mohdsiddique** [comment]
Hi @shirshanka 
I updated the code as per review comments.

**Review by shirshanka — APPROVED** [review]
LGTM!

**Pull request merged**